File: crypto.java
class Constants {
    static final int PUBLIC_KEY_LENGTH = 32;
    static final int PRIVATE_KEY_LENGTH = 32;
    static final int SIGNATURE_LENGTH = 64;

    static final byte MAIN_NET_CHAIN_ID = (byte) 87;
    static final byte TEST_NET_CHAIN_ID = (byte) 84;
}

enum ChainId {
    MAINNET((byte) 87),
    TESTNET((byte) 84);

    ChainId(byte chainId) {
    }
}

//TODO INTERFACES -> CLASSES (but compare with another langs implementations)
//TODO Rename ...Or... args

interface ISeedWithNonce {
    Bytes seed();
    int nonce();
}

/* Type aliases used to increase flexibility and be able
   to extend these types later on. Also type aliases allows
   names to be more self explanatory like in BASE58 case. */

//TODO Bytes Uint8Array

//TODO Base64 = string
//TODO Base58 = string
//TODO Base16 = string

//TODO ChainId = string | number

//TODO Address = Base58
interface IAddress {
}

//TODO Key = Base58 ???

//TODO PublicKey = Key
interface IPublicKey {
}

//TODO PrivateKey = Key
interface IPrivateKey {
}

interface IKeyPair {
    IPublicKey publicKey();
    IPrivateKey privateKey();
}

//TODO Seed = string | Bytes | SeedWithNonce

interface IWavesCrypto {
    ISeedWithNonce seedWithNonce(IBytes seed, int nonce);

    IBytes blake2b(IBytes input);
    IBytes keccak(IBytes input);
    IBytes sha256(IBytes input);

    IBase64 base64encode(IBytes input);
    IBytes base64decode(IBase64 input); //TODO throws (invalid input)
    IBase58 base58encode(IBytes input);
    IBytes base58decode(IBase58 input); //TODO throws (invalid input)
    IBase16 base16encode(IBytes input);
    IBytes base16decode(IBase16 input); //TODO throws (invalid input)

    IKeyPair keyPair(IBytes seed);
    IPublicKey publicKey(IBytes seed);
    IPublicKey publicKey(IPrivateKey privateKey);
    IPrivateKey privateKey(String seed);
    Address address(IPublicKey publicKey, ChainId chainId); //TODO Seed | PrivateKey | PublicKey seedOrKeys, ChainId chainId

    Bytes randomBytes(int size);
    Seed randomSeed();

    Bytes signBytes(Bytes bytes, Seed seed);
    Bytes signBytes(Bytes bytes, PrivateKey privateKey);

    boolean verifySignature(Bytes Bytes, Bytes signature, PublicKey publicKey);
    boolean verifyPublicKey(PublicKey publicKey);
    boolean verifyAddress(Address address, ChainId chainId, PublicKey publicKey); //TODO optional?: { chainId?: TChainId, publicKey?: TPublicKey }
}

File: crypto.kt
typealias Bytes = ByteArray
typealias PublicKey = String
typealias PrivateKey = String
typealias Seed = String
typealias Address = String

const val PUBLIC_KEY_LENGTH = 32
const val PRIVATE_KEY_LENGTH = 32
const val SIGNATURE_LENGTH = 64

const val MAIN_NET_CHAIN_ID = 87
const val TEST_NET_CHAIN_ID = 84

interface KeyPair {
    val publicKey: PublicKey
    val privateKey: PrivateKey
}

interface SeedWithNonce {
  val seed: Seed
  val nonce: Int
}

interface WavesCrypto {

    fun seedWithNonce(seed: Seed, nonce: Int): SeedWithNonce

    fun blake2b(input: Bytes): Bytes
    fun keccak(input: Bytes): Bytes
    fun sha256(input: Bytes): Bytes

    @Throws(ParseException::class)
    fun base58Decode(input: String): Bytes
    fun base58Encode(input: Bytes): String

    @Throws(ParseException::class)
    fun base64Decode(input: String): Bytes
    fun base64Encode(input: Bytes): String

    @Throws(ParseException::class)
    fun base16Decode(input: String): Bytes
    fun base16Encode(input: Bytes): String

    fun keyPair(seed: Seed): KeyPair
    fun keyPair(seed: SeedWithNonce): KeyPair
    fun publicKey(seed: Seed): PublicKey
    fun publicKey(seed: SeedWithNonce): PublicKey
    fun privateKey(seed: Seed): PrivateKey
    fun privateKey(seed: SeedWithNonce): PrivateKey
    
    fun address(publicKey: PublicKey, chainId: String? = null): Address
    fun address(seed: SeedWithNonce, chainId: String? = null): Address
    fun address(seed: Seed, chainId: String? = null): Address

    fun randomSeed(): Seed
    fun randomBytes(size: Int): Bytes

    fun signBytes(bytes: Bytes, privateKey: PrivateKey): Bytes
    fun signBytes(bytes: Bytes, seed: Seed): Bytes

    fun verifySignature(publicKey: PublicKey, bytes: Bytes, signature: Bytes): Boolean
    fun verifyPublicKey(publicKey: PublicKey): Boolean
    fun verifyAddress(address: Address, chainId: String? = null, publicKey: PublicKey? = null): Boolean

}
